### PR TITLE
feat: Azure DevOps Services OAuth2 support

### PR DESCRIPTION
**Pull request by tolusha**
### What does this PR do?
feat: Azure DevOps Services OAuth2 support

### Screenshot/screencast of this PR

N/A

### What issues does this PR fix or reference?

https://issues.redhat.com/browse/CRW-3927

### How to test this PR?

Prerequisite: 
* Deploy Eclipse Che:
  -  `chectl server:deploy --platform <minikube|openshift> --che-operator-image=quay.io/abazko/operator:3927 --cheimage=quay.io/abazko/che-server:3927`
* Set up Azure DevOps Services 
  - Create account for Azure DevOps https://azure.microsoft.com/en-us/products/devops
  - Set the name of your organization and click `Continue`
  - Set the name of your project (it is not the same as repository name) and click `Create project`
  - Click `Repos`, then `Initialize` to create a sample repository
  - Upload any devfile to the repository
    
### OAuth2 workflow
 1. On Azure DevOps service: Click on you Organization name, then `Organization settings` -> `Policies` and turn on `Third-party application access via OAuth`
  - Go to https://aex.dev.azure.com/app/register/ to register application
    - Company name: Che
    - Application name: Che
    - Application website: <CHE_URL>
    - Authorization callback URL: <CHE_URL>/api/oauth/callback
    - Authorization scopes: Code (read and write)
2. Create OAuth secret in Eclipse Che namespace to test :
```yaml
kind: Secret
apiVersion: v1
metadata:
  name: azure-devops-oauth-config
  namespace: eclipse-che
  labels:
    app.kubernetes.io/part-of: che.eclipse.org
    app.kubernetes.io/component: oauth-scm-configuration
  annotations:
    che.eclipse.org/oauth-scm-server: azure-devops
type: Opaque
data:
  id: <base64 encoded App ID>
  secret: <base64 encoded Client Secret>
 ```
 3. Navigate to your repository and click `Clone` and copy the url.
 4. Insert into `Git Repo URL` in Eclipses Che Dasbhoard and click `Create and Open`
 
### PAT workflow:
0. Log out of Eclipse Che,  remove OAuth secret, delete user's namespace, then Log into Eclipse Che again
1. On Azure DevOps service: Click `Personal access token` ->  `New token`
2. Set name, scope: `Code (read and write)` and click `Create`
3. Copy the token
4. Run request to get `scm-user-id`
`curl -s -u  :<PAT> https://vssps.dev.azure.com/<ORGANIZATION>/_apis/connectionData\?api-version\=7.0-preview.1 | jq -r '.authenticatedUser.id'`
5. Create PAT secret (see details https://www.eclipse.org/che/docs/stable/end-user-guide/using-a-git-provider-access-token/):
```yaml
kind: Secret
apiVersion: v1
metadata:
  name: personal-access-token-azure
  namespace: <USER_NAMESPACE>
  labels:
    app.kubernetes.io/component: scm-personal-access-token
    app.kubernetes.io/part-of: che.eclipse.org
  annotations:
    che.eclipse.org/che-userid: <che-user-id>
    che.eclipse.org/scm-personal-access-token-name: azure-devops
    che.eclipse.org/scm-url: https://dev.azure.com/
    che.eclipse.org/scm-userid: <your-user-id>
    che.eclipse.org/scm-username: <your-display-name>
    che.eclipse.org/scm-organization: <your-organization>
data:
  token: <base64 encoded token>
type: Opaque
```
6. Navigate to your repository and click `Clone` and copy the url.
7. Insert into `Git Repo URL` in Eclipses Che Dasbhoard and click `Create and Open`
 
### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/UrlUtils.java [hunk 1]
```diff
@@ -18,10 +18,8 @@
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.net.URLDecoder;
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
```
Comment (vinokurig) [inline core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/UrlUtils.java:22]: Please avoid using wildcard

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 2]
```diff
@@ -113,6 +113,33 @@ che.auth.access_denied_error_page=/error-oauth
...
+che.oauth.azure.devops.authuri=https://app.vssps.visualstudio.com/oauth2/authorize
+
+# Azure DevOps OAuth Service token URI.
+che.oauth.azure.devops.tokenuri=https://app.vssps.visualstudio.com/oauth2/token
+
+# Azure DevOps Service API server address.
+che.integration.azure.devops.api_endpoint=https://vssps.dev.azure.com/
+
+# Azure DevOps SCM API server address.
+che.integration.azure.devops.scm.api_endpoint=https://dev.azure.com/
```
Comment (vinokurig) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:138]: `che.integration.azure.devops.endpoint=https://dev.azure.com` ?

File: wsmaster/che-core-api-factory-azure-devops/src/main/java/org/eclipse/che/api/factory/server/azure/devops/AzureDevOpsAuthorizingFileContentProvider.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    try {
+      urlFetcher.fetch(remoteFactoryUrl.getRepositoryLocation());
+      return true;
+    } catch (IOException e) {
+      return false;
+    }
+  }
+
+  @Override
+  protected String formatAuthorization(String token, boolean isPAT) {
```
Comment (vinokurig) [inline wsmaster/che-core-api-factory-azure-devops/src/main/java/org/eclipse/che/api/factory/server/azure/devops/AzureDevOpsAuthorizingFileContentProvider.java:47]: We are going to have a conflict when https://github.com/eclipse-che/che-server/pull/451 is merged. I prefer to have a separate method for `Basic` authentication.

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java [hunk 4]
```diff
@@ -140,7 +147,14 @@ protected String formatUrl(String fileURL) throws DevfileException {
...
 
-  protected String formatAuthorization(String token) {
+  /**
+   * Formats authorization header value.
+   *
+   * @param token personal access token
+   * @param isPAT true if token is personal access token, false if it is OAuth token
+   * @return formatted authorization header value
+   */
+  protected String formatAuthorization(String token, boolean isPAT) {
```
Comment (vinokurig) [inline wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java:157]: We are going to have a conflict when https://github.com/eclipse-che/che-server/pull/451 is merged. I prefer to have a separate method for `Basic` authentication.

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/PersonalAccessToken.java [hunk 5]
```diff
@@ -22,6 +22,9 @@ public class PersonalAccessToken {
 
   private final String scmProviderUrl;
   private final String scmUserName;
+  /** Organization that user belongs to. Can be null if user is not a member of any organization. */
+  private final String scmOrganization;
```
Comment (vinokurig) [inline wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/PersonalAccessToken.java:26]: Please add `Nullable` annotation

File: wsmaster/che-core-api-factory-azure-devops/src/main/java/org/eclipse/che/api/factory/server/azure/devops/AzureDevOpsApiClient.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+          try {
+            return OBJECT_MAPPER.readValue(response.body(), AzureDevOpsUser.class);
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          }
+        });
+  }
+
+  /** Returns the scopes of the OAuth token. */
+  public String[] getTokenScopes(String authenticationToken)
```
Comment (vinokurig) [inline wsmaster/che-core-api-factory-azure-devops/src/main/java/org/eclipse/che/api/factory/server/azure/devops/AzureDevOpsApiClient.java:130]: The `authenticationToken` parameter is redundant

**Comment by l0rd** [comment]
I have tested on a fresh 4.12 OCP cluster and it worked fine 👍 
A couple of minor comments:
- in the Che dashboard  `User Preferences -> Git services` there is a new line but the "Name" column is empty
- I was expecting `Code(read and write)` to be enough (I haven't tested though)

File: wsmaster/che-core-api-factory-azure-devops/src/main/java/org/eclipse/che/api/factory/server/azure/devops/AzureDevOpsApiClient.java (hunk 6 above)
Comment (tolusha) [reply wsmaster/che-core-api-factory-azure-devops/src/main/java/org/eclipse/che/api/factory/server/azure/devops/AzureDevOpsApiClient.java:130]: Let's keep it for the case I found the way to get the scopes.

Comment (vinokurig) [reply wsmaster/che-core-api-factory-azure-devops/src/main/java/org/eclipse/che/api/factory/server/azure/devops/AzureDevOpsApiClient.java:130]: Please update the javadoc then.

**Comment by tolusha** [comment]
@l0rd 
* https://github.com/eclipse-che/che-dashboard/pull/739
* updated

File: wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketAuthorizingFileContentProvider.java [hunk 7]
```diff
@@ -42,7 +42,7 @@ class BitbucketAuthorizingFileContentProvider extends AuthorizingFileContentProv
 
   /** Formats OAuth token as HTTP Authorization header. */
   @Override
-  protected String formatAuthorization(String token) {
+  protected String formatAuthorization(String token, boolean isPAT) {
```
Comment (ibuziuk) [inline wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketAuthorizingFileContentProvider.java:45]: a bit lost, how isPAT is used?

**Review by ibuziuk — APPROVED** [review]

**Pull request merged**